frame/uniswap-v2/src/lib.rs

```diff
@@ -0,0 +1,857 @@
+		/// - pool identifier `T::PoolId`
+		/// - added token amounts `Vec<T::Balance>`
+		/// - charged fees `Vec<T::Balance>`
+		/// - actual invariant `T::Balance`
+		/// - actual token supply `T::Balance`
+		/// - minted amount `T::Balance`
+		///
+		/// \[who, pool_id, token_amounts, fees, invariant, token_supply, mint_amount\]
+		LiquidityAdded {
+			who: T::AccountId,
+			pool_id: T::PoolId,
+			token_amounts: Vec<T::Balance>,
+			fees: Vec<T::Balance>,
+			invariant: T::Balance,
+			token_supply: T::Balance,
+			mint_amount: T::Balance,
+		},
+
+		/// Liquidity removed from pool `T::PoolId` by `T::AccountId` in balanced way.
+		///
+		/// Included values are:
+		/// - account identifier `T::AccountId`
+		/// - pool identifier `T::PoolId`
+		/// - removed token amounts `Vec<T::Balance>`
+		/// - charged fees `Vec<T::Balance>`
+		/// - actual token supply `T::Balance`
+		///
+		/// \[who, pool_id, token_amounts, fees, token_supply\]
+		LiquidityRemoved {
+			who: T::AccountId,
+			pool_id: T::PoolId,
+			token_amounts: Vec<T::Balance>,
+			fees: Vec<T::Balance>,
+			token_supply: T::Balance,
+		},
+
+		/// Token exchange happened.
+		///
+		/// Included values are:
+		/// - account identifier `T::AccountId`
+		/// - pool identifier `T::PoolId`
+		/// - index of sent token `T::PoolTokenIndex`
+		/// - amount of sent token `T::Balance`
+		/// - index of received token `T::PoolTokenIndex`
+		/// - amount of received token `T::Balance`
+		/// - charged fee `T::Balance`
+		///
+		/// \[who, pool_id, sent_token_index, sent_amount, received_token_index, received_amount,
+		/// fee\]
+		TokenExchanged {
+			who: T::AccountId,
+			pool_id: T::PoolId,
+			sent_token_index: T::PoolTokenIndex,
+			sent_amount: T::Balance,
+			received_token_index: T::PoolTokenIndex,
+			received_amount: T::Balance,
+			fee: T::Balance,
+		},
+
+		/// Withdraw admin fees `Vec<T::Balance>` from pool `T::PoolId` by user `T::AccountId`
+		///
+		/// Included values are:
+		/// - account identifier `T::AccountId`
+		/// - pool identifier `T::PoolId`
+		/// - admin fee receiving account identifier `T::AccountId`
+		/// - withdrew admin fees `Vec<T::Balance>`
+		///
+		/// [who, pool_id, admin_fee_account, admin_fees]
+		AdminFeesWithdrawn {
+			who: T::AccountId,
+			pool_id: T::PoolId,
+			admin_fee_account: T::AccountId,
+			admin_fees: Vec<T::Balance>,
+		},
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {}
+
+	impl<T: Config> CurveAmm for Pallet<T> {
+		type AssetId = T::AssetId;
+		type Balance = T::Balance;
+		type AccountId = T::AccountId;
+		type PoolId = T::PoolId;
+		type PoolTokenIndex = T::PoolTokenIndex;
+
+		fn pool_count() -> T::PoolId {
+			PoolCount::<T>::get()
+		}
+
+		fn add_liquidity(
+			who: &Self::AccountId,
+			pool_id: T::PoolId,
+			amounts: Vec<Self::Balance>,
+			min_mint_amount: Self::Balance,
+		) -> Result<(), DispatchError> {
+			let zero = Self::Balance::zero();
+			ensure!(amounts.len() == 2, Error::<T>::IndexOutOfRange);
```

Comment:
not clear what index out of range means. it is seems like you encode Exact2Values type via array. can we encode it into types than.

most of swaps are in pools of 2, not in many. so having separate interface will be more clear and reusable. also it will fix potential security issuew ith Vec

Reply:
Agreed , will remove Vec in next PR and hopefully it will address this concern.